1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+This is proposed because:
+
+- It builds off an existing PowerShell concept
+- It applies to both native commands and cmdlets/functions
+- Diverging from this would create inconsistency
+- Changes to the behaviour of `$?` would be reasonably expected to change here
+
+That is, there are the following direct equivalents:
+
+```none
+cmd1 && cmd2
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 })
+```
+
+```none
+cmd1 || cmd2
+
+    |
+    v
+
+$(cmd1; if (-not $?) { cmd2 })
+```
+
+```none
+cmd1 && cmd2 || cmd3
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 }; if (-not $?) { cmd3 })
+# Note that cmd1 failing runs cmd3
+```
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+#### Pipeline output
+
+The output of a pipeline chain is the concatenation of all its pipelines,
+in the sequence of their output:
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Output 'd'
+$x # 'a','b','c','d'
+```
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Error 'Bad' || Write-Output 'd'
+# Writes the error record 'Bad'
+$x # 'a','b','c','d'
+```
+
+Commands that fail will have any pipeline output
+emitted before evaluation of the chain operator.
+
+For example:
+
+```powershell
+$x = cmd1 && cmd2
+```
+
+If `cmd1` fails but emits output, `$x` will hold that value.
+
+#### Error semantics
+
+Errors will have the same semantics as the equivalent
+`cmd1; if ($?) { cmd2 }` syntax.
+
+Terminating errors will terminate the entire pipeline chain.
+
+Non-terminating errors will cause the immediate pipeline to continue,
+and the pipeline chain will evaluate as normal based on the value of `$?`.
+
+### Other notes
+
+#### New Abstract Syntax Tree (AST) types
+
+The new AST type, `PipelineChainAst`,
+will be created to represent a pipeline chain.
+This will inherit from `PipelineBaseAst`
+and may occur anywhere a `PipelineBaseAst` does in a PowerShell AST.
+
+`ICustomAstVisitor2` and `AstVisitor2` would be extended to deal with this AST,
+and .NET Core 3's new default interface implementation feature would be
+leveraged to ensure this does not break things
+as previous syntactic introductions have been forced to.
+
+#### Statements may not be chained
+
+PowerShell has a notable divergence from POSIX shells:
+
+- In POSIX shells, a pipeline is composed of statements
+- In PowerShell, a pipeline is a kind of statement
+
+This means that bash allows the following:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF'
+```
+
+In fact, this can be evaluated in the background:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF' &
+```
+
+However, in PowerShell, pipelines are subordinate to statements.
+The proposed implementation of chain operators precludes use of
+`&&` and `||` between statements like `if` or `while`,
+or constructions like `$x = cmd1 && $y = cmd2 && $x + $y`.
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+

```

Comment:
Following some of the earlier comments on how to store results of different pipelines in different variables (which I believe will be an important scenario to support), and the comments about our precedence differing from bash, I'm leaning towards feeling this would be better as a statement chain than as a pipeline chain.